src/sampling.jl

```diff
@@ -0,0 +1,177 @@
+
+"""Iterator interface functions for RejectionSampler"""
+Base.iterate(sampler::RejectionSampler, ::Nothing) = iterate(sampler)
+
+function Base.iterate(sampler::RejectionSampler)
+    if sampler.accepted >= sampler.num_samples
+        return nothing
+    else
+        return prod(rand(sampler.rng, ["0", "1"], sampler.num_qubits)), nothing
+    end
+end
+
+"""
+    accept!(results::Samples{T}, sampler::RejectionSampler, bitstring::String) where T<:Complex
+
+Accept or reject the given bitstring as a sample using the rejection method and update 
+`results` accordingly.
+"""
+function accept!(results::Samples{T}, sampler::RejectionSampler, bitstring::String) where T<:Complex
+    # Get the amplitude for the given bitstring and the parameters for the rejection method.
+    amp = results.amplitudes[bitstring]
+    Np = 2^sampler.num_qubits * abs(amp)^2
+    sampler.fix_M && (sampler.M = max(Np, sampler.M))
+
+    # Accept the given bitstring as a sample with probability Np/M.
+    u = rand(sampler.rng)
+    if u < Np / sampler.M
+        sampler.accepted += 1
+        results.bitstrings[bitstring] += 1
+    end
+end
+
+###############################################################################
+# UniformSampler
+###############################################################################
+
+"""
+A Sampler struct to uniformly sample bitstrings and compute their amplitudes.
+"""
+mutable struct UniformSampler <: AbstractSampler
+    num_qubits::Int64
+    num_samples::Int64
+    rng::MersenneTwister
+end
+
+function UniformSampler(params)
+    num_qubits = params["num_qubits"]
+    num_samples = params["num_samples"]
+    rng = MersenneTwister(params["seed"])
+    UniformSampler(num_qubits, num_samples, rng)
+end
+
+"""Iterator interface functions for UniformSampler"""
+Base.iterate(sampler::UniformSampler) = iterate(sampler, 0)
+
+function Base.iterate(sampler::UniformSampler, samples_produced::Integer)
+    if samples_produced < sampler.num_samples
+        new_bitstring = prod(rand(sampler.rng, ["0", "1"], sampler.num_qubits))
+        return new_bitstring, samples_produced + 1
+    else
+        return nothing
+    end
+end
+
+"""
+    accept!(results::Samples{T}, ::UniformSampler, bitstring::String) where T<:Complex
+
+Accept the given bitstring as a sample and update `results` accordingly.
+"""
+function accept!(results::Samples{T}, ::UniformSampler, bitstring::String) where T<:Complex
+    results.bitstrings[bitstring] += 1
+end
+
+###############################################################################
+# Sampler Constructor
+###############################################################################
+
+CONSTRUCTORS = Base.ImmutableDict(
+    "rejection" => RejectionSampler,
+    "uniform" => UniformSampler,
+    "list" => ListSampler
+)
+
+"""
+    construct_sampler(params)
+
+Returns a sampler whose type is specified in `params`.
+"""
+function construct_sampler(params)
+    CONSTRUCTORS[params["output_method"]](params)
```

Comment:
This patter of using a dict that points to functions can  be a little clunky as have to maintain this lookup table as well as functions. Can go directly from a string to a function so if the "output_method" was the name of the function could do something like

```
get_func(func_name::String) = getfield(Main, Symbol(func_name))
```
and can then do

```
get_func(params["output_method"])(params)
```